### ISSUE TITLE
Clearly document the limitations with known_hosts

### ISSUE THREAD

**Issue by asarkar** [spring-cloud/spring-cloud-config#623]
In order to successfully connect to a repo using ssh, among other things, 2 things are required:
1. An entry for the Git server must be present in the `~/.ssh/known_hosts` file.
2. The entry must be in the `ssh-rsa` format, not `ecdsa-sha2-nistp256`.

Either one of the above steps missing results in a NPE.

```
java.lang.NullPointerException: null
	at com.jcraft.jsch.Session.checkHost(Session.java:802) ~[jsch-0.1.50.jar:na]
	at com.jcraft.jsch.Session.connect(Session.java:342) ~[jsch-0.1.50.jar:na]
	at org.eclipse.jgit.transport.JschConfigSessionFactory.getSession(JschConfigSessionFactory.java:116) ~[org.eclipse.jgit-3.5.3.201412180710-r.jar:3.5.3.201412180710-r]
	at org.eclipse.jgit.transport.SshTransport.getSession(SshTransport.java:136) ~[org.eclipse.jgit-3.5.3.201412180710-r.jar:3.5.3.201412180710-r]
	at org.eclipse.jgit.transport.TransportGitSsh$SshFetchConnection.<init>(TransportGitSsh.java:262) ~[org.eclipse.jgit-3.5.3.201412180710-r.jar:3.5.3.201412180710-r]
	at org.eclipse.jgit.transport.TransportGitSsh.openFetch(TransportGitSsh.java:161) ~[org.eclipse.jgit-3.5.3.201412180710-r.jar:3.5.3.201412180710-r]
	at org.eclipse.jgit.transport.FetchProcess.executeImp(FetchProcess.java:136) ~[org.eclipse.jgit-3.5.3.201412180710-r.jar:3.5.3.201412180710-r]
	at org.eclipse.jgit.transport.FetchProcess.execute(FetchProcess.java:122) ~[org.eclipse.jgit-3.5.3.201412180710-r.jar:3.5.3.201412180710-r]
	at org.eclipse.jgit.transport.Transport.fetch(Transport.java:1115) ~[org.eclipse.jgit-3.5.3.201412180710-r.jar:3.5.3.201412180710-r]
	at org.eclipse.jgit.api.FetchCommand.call(FetchCommand.java:130) ~[org.eclipse.jgit-3.5.3.201412180710-r.jar:3.5.3.201412180710-r]
	at org.eclipse.jgit.api.CloneCommand.fetch(CloneCommand.java:178) ~[org.eclipse.jgit-3.5.3.201412180710-r.jar:3.5.3.201412180710-r]
	at org.eclipse.jgit.api.CloneCommand.call(CloneCommand.java:125) ~[org.eclipse.jgit-3.5.3.201412180710-r.jar:3.5.3.201412180710-r]

```

JGit hasn't realized that the year is 2017 and still use a mailing list to report bugs which is more time than I'm willing to spend (do they also use Yahoo chat for Dev communication?). Thus, I think, Spring Cloud should document this clearly and also suggest a fix. The fix is to remove any existing entries from `known_hosts` that's using `ecdsa-sha2-nistp256` for the Git server and put in a new entry in `ssh-rsa` format by running the command below:
`ssh-keyscan -t rsa GIT_SERVER >> ~/.ssh/known_hosts`

**Comment by ryanjbaxter** [comment]
@asarkar aren't both points already called out in the documentation?

> If you don’t use HTTPS and user credentials, SSH should also work out of the box when you store keys in the default directories (~/.ssh) and the uri points to an SSH location, e.g. "git@github.com:configuration/cloud-configuration". It is important that all keys in ~/.ssh/known_hosts are in "ssh-rsa" format. The new "ecdsa-sha2-nistp256" format is NOT supported.

Are there some changes you think we should make to this?  If so would you like to submit a PR for them?

**Comment by asarkar** [comment]
@ryanjbaxter I think the section you quoted specifies the format of the keys but doesn't say that those must be present in the known_hosts file, or if there is a way to bypass this; perhaps there is a property you can set to automatically accept the key; it won't be the most secure system, but less hassle for Dev. For a cmd line ssh client, having the key previously present in the known_hosts file is not a requirement; the clients asks you when you attempt to make a connection to a new server. This can even be bypassed by passing `oStrictHostKeyChecking=no`.
NPE is not acceptable in any case because then the user, me, has to debug someone else's code to find out what they're looking for. Throwing an exception with a relevant error message is ok too.

**Comment by ryanjbaxter** [comment]
> @ryanjbaxter I think the section you quoted specifies the format of the keys but doesn't say that those must be present in the known_hosts file

I think that is what is being said in this sentence.

> It is important that all keys in ~/.ssh/known_hosts are in "ssh-rsa" format.

Perhaps it could be stated more explicitly?

I agree that the `NPE` is not very dev friendly but it is not Spring Cloud that is throwing the `NPE` it is JGit (at least that is what I can see from the stack trace you provided), so I am not so sure what we can do about it.  If you think that the documentation can be enhanced to help out, we are open to suggestions.

**Comment by asarkar** [comment]
@ryanjbaxter I'll submit a PR.

**Comment by asarkar** [comment]
https://github.com/spring-cloud/spring-cloud-config/pull/632

**Comment by izeye** [comment]
Resolved via #632 ?

**Comment by ryanjbaxter** [comment]
@izeye yes

@asarkar for future reference, if in the description of your PR you say something like "Fixes #<issue-number>"  GitHub will automatically close the issue when the PR is merged.
